src/test/java/com/diffblue/deeptestutils/mock/MethodMatcherTest.java

```diff
@@ -0,0 +1,277 @@
+  }
+
+  static Parent simple(Parent a) {
+    return new Parent("0");
+  }
+
+  static Parent returnSubtype(Parent a) {
+    return new Parent("2");
+  }
+
+  static Parent argSubtype(Parent a) {
+    return new Parent("4");
+  }
+
+  static Parent argSupertypeReturnSubtype(Child a) {
+    return new Parent("6");
+  }
+
+  static Parent mix1(Child a) {
+    return new Parent("8");
+  }
+}
+
+class Child extends Parent{
+  public Child(String val) {
+     super(val);
+  }
+
+  static Parent simple(Parent a) {
+   return new Parent("1");
+  }
+
+  static Child returnSubtype(Parent a) {
+    return new Child("3");
+  }
+
+  static Parent argSubtype(Child a) {
+    return new Parent("5");
+  }
+
+  static Child argSupertypeReturnSubtype(Parent a) {
+    return new Child("7");
+  }
+
+  static Parent mix1(Child a) {
+    return new Child("9x");
+  }
+
+  static Child mix1(Parent a) {
+    return new Child("9");
+  }
+}
+
+public class MethodMatcherTest {
+
+  @org.junit.Rule
+  public ExpectedException thrown = ExpectedException.none();
+
+  @org.junit.Test
+  public void simpleTest() throws IllegalAccessException, InvocationTargetException {
+    Method m1 = DTUMemberMatcher.method(Child.class, "simple", Parent.class);
+    Method m2 = DTUMemberMatcher.methodUsingWholeSignature(Child.class,"simple", Parent.class);
+
+    Child child = new Child("child");
+    Parent parent = new Parent("parent");
+
+    String s1 = m1.invoke(child, parent).toString();
+    String s2 = m2.invoke(child, parent).toString();
+
+    // verify that we find the same method using both DTUMemberMatcher methods
+    Assert.assertEquals(s1, s2);
+    // verify that we find the correct method according to Java language specification
+    Assert.assertEquals(child.simple(parent).toString(), s1);
+    // verify that the matched method returns what we expect it to return
+    Assert.assertEquals("1", s1);
+  }
+
+  @org.junit.Test
+  public void returnSubtypeTest() throws IllegalAccessException, InvocationTargetException {
+    Method m1 = DTUMemberMatcher.method(Child.class, "returnSubtype", Parent.class);
+    Method m2 = DTUMemberMatcher.methodUsingWholeSignature(Child.class, "returnSubtype", Parent.class);
+
+    Child child = new Child("child");
+    Parent parent = new Parent("parent");
+
+    String s1 = m1.invoke(child, parent).toString();
+    String s2 = m2.invoke(child, parent).toString();
+
+    // verify that we find the same method using both DTUMemberMatcher methods
+    Assert.assertEquals(s1, s2);
+    // verify that we find the correct method according to Java language specification
+    Assert.assertEquals(child.returnSubtype(parent).toString(), s1);
+    // verify that the matched method returns what we expect it to return
+    Assert.assertEquals("3", s1);
+  }
+
+  @org.junit.Test
+  public void argSubtypeTest() throws IllegalAccessException, InvocationTargetException {
+    Method m1 = DTUMemberMatcher.method(Child.class, "argSubtype", Child.class);
```

Comment:
:pick: In some tests the string was defined as a variable, in other tests not. Maybe make it consistent (prefer the variable kind).